### PR TITLE
feat: add cli and api for delete checkpoints [DET-7119], [DET-7120]

### PR DESCRIPTION
**Pull request by nrajanee**
## Description

feat: add functionality and add an api endpoint for deleting a list of checkpoints given their uuids [DET-7119], [DET-7120]

## Test Plan

Tested well in integration and e2e tests.

[DET-7119]: https://determinedai.atlassian.net/browse/DET-7119?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-7120]: https://determinedai.atlassian.net/browse/DET-7120?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: e2e_tests/tests/cluster/test_checkpoints.py [hunk 1]
```diff
@@ -45,6 +48,75 @@ def test_gc_checkpoints_lfs() -> None:
...
+    exp_2_checkpoints = bindings.get_GetExperimentCheckpoints(
+        session=test_session, id=exp_id_1
+    ).checkpoints
+    assert len(exp_1_checkpoints) > 0, f"no checkpoints found in experiment with ID:{exp_id_1}"
+    assert len(exp_2_checkpoints) > 0, f"no checkpoints found in experiment with ID:{exp_id_2}"
+
+    indices_exp_1 = [i for i in range(0, len(exp_1_checkpoints))]
+    indices_exp_2 = [i for i in range(0, len(exp_2_checkpoints))]
+    d_exp_1_checkpoint_uuids = [
+        exp_1_checkpoints[d_index].uuid for d_index in random.sample(indices_exp_1, 2)
```
Comment (stoksc) [inline e2e_tests/tests/cluster/test_checkpoints.py:83]: I think instead you can just do `random.sample(range(len(exp_1_checkpoints)), 2)`

File: e2e_tests/tests/cluster/test_checkpoints.py [hunk 2]
```diff
@@ -45,6 +48,75 @@ def test_gc_checkpoints_lfs() -> None:
...
+    ]
+    d_exp_2_checkpoint_uuids = [
+        exp_2_checkpoints[d_index].uuid for d_index in random.sample(indices_exp_2, 2)
+    ]
+
+    d_checkpoint_uuids = d_exp_1_checkpoint_uuids + d_exp_2_checkpoint_uuids
+    delete_body = bindings.v1DeleteCheckpointsRequest(checkpointUuids=d_checkpoint_uuids)
+    bindings.delete_DeleteCheckpoints(session=test_session, body=delete_body)
+    wait_for_gc_to_finish(exp_id_1)
```
Comment (stoksc) [inline e2e_tests/tests/cluster/test_checkpoints.py:92]: we should probably call this after `run_basic_test_with_temp_config` so we can make sure we're actually checkpoint for the deletion of checkpoints that aren't getting deleted by the normal gc job.

File: e2e_tests/tests/cluster/test_checkpoints.py [hunk 3]
```diff
@@ -45,6 +48,75 @@ def test_gc_checkpoints_lfs() -> None:
...
+def ensure_checkpoint_deleted(test_session, config, d_checkpoint_uuid) -> bool:
+    d_checkpoint = bindings.get_GetCheckpoint(
+        session=test_session, checkpointUuid=d_checkpoint_uuid
+    ).checkpoint
+    if d_checkpoint is not None:
+        assert (
+            d_checkpoint.state == determinedcheckpointv1State.STATE_DELETED
+        ), f"checkpoint with uuid {d_checkpoint_uuid} does not have a deleted state"
+    else:
+        raise Exception(
```
Comment (stoksc) [inline e2e_tests/tests/cluster/test_checkpoints.py:108]: here and elsewhere, maybe a good thing to use `pytest.fail` instead of `raise` to distinguish test failures from test exceptions. though we don't really hold this up all over, i was just thinking about it in tests i was writing recently.

File: harness/determined/cli/checkpoint.py [hunk 4]
```diff
@@ -168,6 +182,20 @@ def describe(args: Namespace) -> None:
             "describe checkpoint",
             [Arg("uuid", type=str, help="checkpoint uuid to describe")],
         ),
+        Cmd(
+            "delete",
+            delete_checkpoints,
+            "delete checkpoints",
+            [
+                Arg("checkpoints_uuids", help="comma-separated list of checkpoints to delete"),
```
Comment (stoksc) [inline harness/determined/cli/checkpoint.py:190]: just thinking, what is better `det c delete 1,2,3` or `det c delete 1 2 3`? thoughts?

File: harness/determined/exec/gc_checkpoints.py [hunk 5]
```diff
@@ -99,8 +101,8 @@ def main(argv: List[str]) -> None:
 
     manager = storage.build(storage_config, container_path=constants.SHARED_FS_CONTAINER_PATH)
 
-    storage_ids = [c["uuid"] for c in args.delete["checkpoints"]]
-
+    logging.info(f"This is args.delete: {args.delete}")
```
Comment (stoksc) [inline harness/determined/exec/gc_checkpoints.py:104]: don't forget to remove

File: harness/determined/exec/gc_checkpoints.py [hunk 6]
```diff
@@ -22,6 +22,8 @@ def delete_checkpoints(
     logging.info("Deleting {} checkpoints".format(len(to_delete)))
 
     for storage_id in to_delete:
+        if storage_id == "":
```
Comment (stoksc) [inline harness/determined/exec/gc_checkpoints.py:25]: why was this necessary? fine as a precaution but in general i would expect this validation to happen earlier

File: master/internal/db/postgres_checkpoints.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
+package db
+
+import (
+	"fmt"
+
+	"github.com/google/uuid"
+)
+
+// FilterForRegisteredCheckpoints gets the deleted checkpoints provided in the model registry.
+func (db *PgDB) FilterForRegisteredCheckpoints(deleteCheckpoints []uuid.UUID) ([]uuid.UUID, error) {
```
Comment (stoksc) [inline master/internal/db/postgres_checkpoints.go:10]: `deleteCheckpoints` -> `checkpoints` or `cs`.. this method has no idea they're being deleted.

Comment (stoksc) [reply master/internal/db/postgres_checkpoints.go:10]: also the comment implies the checkpoints passed are deleted.

File: master/internal/db/postgres_checkpoints.go [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+	var checkpointIDRows []struct {
+		ID uuid.UUID
+	}
+
+	if err := db.queryRows(`
+	SELECT DISTINCT(c.uuid) AS ID FROM checkpoints_view AS c
```
Comment (stoksc) [inline master/internal/db/postgres_checkpoints.go:16]: we don't nede to join at all here, we can just `SELECT DISTINCT(mv.checkpoint_uuid) FROM model_versions AS mv WHERE mv.checkpoint_uuid IN (SELECT UNNEST($1::uuid[]))`.

File: master/internal/db/postgres_checkpoints.go [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+	}
+
+	return checkpointIDs, nil
+}
+
+// MarkCheckpointsDeleted updates the provided delete checkpoints to DELETED state.
+func (db *PgDB) MarkCheckpointsDeleted(deleteCheckpoints []uuid.UUID) error {
+	_, err := db.sql.Exec(`UPDATE raw_checkpoints c
+    SET state = 'DELETED'
+    WHERE c.uuid IN (SELECT UNNEST($1::uuid[]))`, deleteCheckpoints)
```
Comment (stoksc) [inline master/internal/db/postgres_checkpoints.go:37]: we can talk more offline, but let's not use this `unnest` thing going forward. we ran into a really weird perf issue with it in a more complex query.

File: master/internal/db/postgres_checkpoints.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+// GroupCheckpointUUIDsByExperimentID creates the mapping of checkpoint uuids to experiment id.
+// The checkpount uuids grouped together are comma separated.
+func (db *PgDB) GroupCheckpointUUIDsByExperimentID(checkpoints []uuid.UUID) (
+	[]*ExperimentCheckpointGrouping, error) {
+	var groupeIDcUUIDS []*ExperimentCheckpointGrouping
+
+	rows, err := db.sql.Queryx(
+		`SELECT e.id AS ExperimentID, string_agg(c.uuid::text, ',') AS CheckpointUUIDSStr
+	FROM experiments e
+	JOIN checkpoints_view c ON c.experiment_id = e.id
```
Comment (stoksc) [inline master/internal/db/postgres_checkpoints.go:67]: also don't need a join here, `checkpoints_view` has the experiment id.

File: master/internal/db/postgres_checkpoints.go [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+	WHERE c.uuid IN (SELECT UNNEST($1::uuid[]))
+	GROUP BY e.id`, checkpoints)
+	if err != nil {
+		return nil, fmt.Errorf("grouping checkpoint UUIDs by experiment ids: %w", err)
+	}
+
+	for rows.Next() {
```
Comment (stoksc) [inline master/internal/db/postgres_checkpoints.go:74]: just using `queryRows` like above may save all this code

File: master/internal/db/postgres_experiments.go [hunk 12]
```diff
@@ -1220,36 +1221,27 @@ WITH const AS (
...
-                           #- '{experiment_rank}' #- '{trial_rank}' #- '{trial_order_rank}'
-                       ORDER BY id ASC), '[]'::jsonb)
-            FROM selected_checkpoints
-           ) AS checkpoints
-    FROM const
-) x
-`
-
-	return db.rawQuery(ctes+query, id, experimentBest, trialBest, trialLatest)
+	if err := db.queryRows(ctes+query, &checkpointIDRows,
```
Comment (stoksc) [inline master/internal/db/postgres_experiments.go:1230]: no reason for this string addition now, just add them statically

File: master/internal/db/postgres_experiments.go [hunk 13]
```diff
@@ -1140,14 +1142,13 @@ FROM experiments
...
-// checkpoints are also marked as deleted in the database.
+// ExperimentCheckpointsToGCRaw returns a comma-separated string describing checkpoints
+// that should be GCed according to the given GC policy parameters. If the delete parameter is true,
+// the returned checkpoints are also marked as deleted in the database.
 func (db *PgDB) ExperimentCheckpointsToGCRaw(
 	id int,
 	experimentBest, trialBest, trialLatest int,
-	delete bool,
-) ([]byte, error) {
+) (string, error) {
```
Comment (stoksc) [inline master/internal/db/postgres_experiments.go:1151]: bit of a nit but, may as well have this return `[]string` and let another layer do the serialization.

File: master/internal/api_checkpoint.go [hunk 14]
```diff
@@ -28,6 +35,82 @@ func (a *apiServer) GetCheckpoint(
...
 
+func (a *apiServer) DeleteCheckpoints(
+	ctx context.Context,
+	req *apiv1.DeleteCheckpointsRequest) (*apiv1.DeleteCheckpointsResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx, a.m.db, &a.m.config.InternalConfig.ExternalSessions)
+	if err != nil {
+		return nil, err
+	}
+
+	checkpointsToDeleteStr := req.CheckpointUuids
```
Comment (stoksc) [inline master/internal/api_checkpoint.go:46]: this temp var is longer than `req.CheckpointUuids` and only used once, so let's just get rid of it

File: master/internal/db/postgres_checkpoints.go [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+		ID uuid.UUID
+	}
+
+	if err := db.queryRows(`
+	SELECT DISTINCT(c.uuid) AS ID FROM checkpoints_view AS c 
+	JOIN model_versions AS mv ON mv.checkpoint_uuid = c.uuid
+	WHERE c.uuid IN (SELECT UNNEST($1::uuid[])); 
+`, &checkpointIDRows, deleteCheckpoints); err != nil {
+		return nil, fmt.Errorf(
+			"querying for all requested delete checkpoints registered in model registry: %w", err)
```
Comment (stoksc) [inline master/internal/db/postgres_checkpoints.go:21]: the error is describing the callers context - let them describe that.
`fmt.Errorf("filtering checkpoint uuids (%v) by those registered in the model registry: %w", uuids, err)`

File: master/internal/api_checkpoint.go [hunk 16]
```diff
@@ -28,6 +35,82 @@ func (a *apiServer) GetCheckpoint(
...
+
+	registeredCheckpointUUIDs, err := a.m.db.FilterForRegisteredCheckpoints(checkpointsToDelete)
+	if err != nil {
+		return nil, err
+	}
+
+	// return 400 if model registry checkpoints and include all the model registry checkpoints
+	if len(registeredCheckpointUUIDs) > 0 {
+		return nil, status.Errorf(codes.InvalidArgument,
+			"This subset of checkpoints provided are in the model registry and cannot be deleted: %v.",
```
Comment (stoksc) [inline master/internal/api_checkpoint.go:62]: maybe lower case, since we're sort of consistent about that elsewhere (even right above)

File: master/internal/api_checkpoint.go [hunk 17]
```diff
@@ -28,6 +35,82 @@ func (a *apiServer) GetCheckpoint(
...
+			registeredCheckpointUUIDs)
+	}
+
+	addr := actor.Addr(fmt.Sprintf("checkpoints-gc-%s", uuid.New().String()))
+
+	taskSpec := *a.m.taskSpec
+	agentUserGroup, err := a.m.db.AgentUserGroup(curUser.ID)
```
Comment (stoksc) [inline master/internal/api_checkpoint.go:69]: todo for us: make all the defaulting logic (like agentUserGroup here) way clearer and harder to mess  up

File: master/internal/api_checkpoint.go [hunk 18]
```diff
@@ -28,6 +35,82 @@ func (a *apiServer) GetCheckpoint(
...
+	for _, expIDcUUIDs := range groupCUUIDsByEIDs {
+		exp, eErr := a.m.db.ExperimentByID(expIDcUUIDs.ExperimentID)
+		if eErr != nil {
+			return nil, err
+		}
+
+		jobSubmissionTime := time.Now().UTC().Truncate(time.Millisecond)
+		ckptGCTask := newCheckpointGCTask(a.m, jobID, jobSubmissionTime, taskSpec,
+			exp.ID, exp.Config.AsLegacy(), expIDcUUIDs.CheckpointUUIDSStr, agentUserGroup, curUser)
+		if gcErr := a.m.system.MustActorOf(addr,
```
Comment (stoksc) [inline master/internal/api_checkpoint.go:105]: `gcErr` isn't an error, it's an actor ref that should never be nil (else we would panic, hence the "must" in the name). so i'm suprisied this doesn't always error?

File: master/internal/checkpoint_gc.go [hunk 19]
```diff
@@ -33,6 +36,28 @@ type checkpointGCTask struct {
 	logCtx logger.Context
 }
 
+func newCheckpointGCTask(master *Master, jobID model.JobID, jobSubmissionTime time.Time,
```
Comment (stoksc) [inline master/internal/checkpoint_gc.go:39]: let's not pass this a ref to the master. instead just pass the db and rm

File: master/internal/checkpoint_gc.go [hunk 20]
```diff
@@ -77,17 +102,31 @@ func (t *checkpointGCTask) Receive(ctx *actor.Context) error {
...
 		if msg.Child.Address().Local() == t.allocationID.String() {
 			t.completeTask(ctx)
+			return errors.Wrapf(msg.Error, "checkpoint GC task failed (actor.ChildFailed)")
 		}
 	case actor.PostStop:
 	default:
 		return actor.ErrUnexpectedMessage(ctx)
 	}
+
+	conv := &protoconverter.ProtoConverter{}
```
Comment (stoksc) [inline master/internal/checkpoint_gc.go:112]: this will run on the first message that gets through the switch case (so `actor.PreStart`) which isn't right. we want to run this only once the allocation actor we kick off to run the GC job has exited successfully, which we will know when we receive the `task.AllocationExited` and it has a nil error

File: master/internal/experiment.go [hunk 21]
```diff
@@ -345,7 +345,6 @@ func (e *experiment) Receive(ctx *actor.Context) error {
 			e.Config.CheckpointStorage().SaveExperimentBest(),
 			e.Config.CheckpointStorage().SaveTrialBest(),
 			e.Config.CheckpointStorage().SaveTrialLatest(),
-			true,
 		)
 		if err != nil {
 			ctx.Log().WithError(err).Error("")
```
Comment (stoksc) [inline master/internal/experiment.go:350]: we should use `newCheckpointGCTask` below, too (we may need to change the singature a tad).

File: master/static/migrations/20220512160917_update-job-type-jobs.tx.down.sql [hunk 22]
```diff
@@ -0,0 +1,10 @@
+ALTER TYPE public.job_type RENAME TO _job_type; 
+CREATE TYPE public.job_type AS ENUM (
+    'EXPERIMENT',
+    'NOTEBOOK',
+    'SHELL',
+    'COMMAND',
+    'TENSORBOARD'
+);
+ALTER TABLE public.jobs ALTER COLUMN job_type TYPE public.job_type USING (job_type::text::job_type);
+DROP TYPE _job_type;
```
Comment (stoksc) [inline master/static/migrations/20220512160917_update-job-type-jobs.tx.down.sql:10]: this migration isn't necessarily safe to run if we ran on this version for a while. probably just run a `DELETE FROM public.jobs WHERE job_type = 'CHECKPOINT_GC'` first or something.

**Review by stoksc — COMMENTED** [review]
looking very good, nearly there. thanks for all the hard work on this.

File: harness/determined/cli/checkpoint.py (hunk 4 above)
Comment (nrajanee) [reply harness/determined/cli/checkpoint.py:190]: I think 1,2,3 looks more like a list than 1 2 3. But I don't think I have strong preference. What do most cli's do? I saw that in experiment.py (describe experiments) we use commas.

File: e2e_tests/tests/cluster/test_checkpoints.py (hunk 2 above)
Comment (nrajanee) [reply e2e_tests/tests/cluster/test_checkpoints.py:92]: (messaged on slack about this).

File: harness/determined/exec/gc_checkpoints.py (hunk 6 above)
Comment (nrajanee) [reply harness/determined/exec/gc_checkpoints.py:25]: true! will add a precaution earlier in cli/checkpoints.py to check if it's a valid uuid.

File: master/internal/db/postgres_checkpoints.go (hunk 8 above)
Comment (nrajanee) [reply master/internal/db/postgres_checkpoints.go:16]: oops yes! not sure why i used join, thank you!

File: master/internal/db/postgres_checkpoints.go (hunk 11 above)
Comment (nrajanee) [reply master/internal/db/postgres_checkpoints.go:74]: for queryRows in the struct I can only use identifiers such as ID. I think they need to be defined in sql? That's why I used StructScan cause then I can use an identifier that is appropriate for the values.

File: master/internal/db/postgres_experiments.go (hunk 13 above)
Comment (nrajanee) [reply master/internal/db/postgres_experiments.go:1151]: then every method that calls this method would have to do the serialization? Isn't it better to do it here?

File: master/internal/api_checkpoint.go (hunk 18 above)
Comment (nrajanee) [reply master/internal/api_checkpoint.go:105]: I'm not sure about this. I will look into it. Even in api_experiment.go (line 215) it doesn't seem to error?

Comment (stoksc) [reply master/internal/api_checkpoint.go:104]: oh wait, duh, it ends with .AwaitTermination

Comment (stoksc) [reply master/internal/api_checkpoint.go:104]: if you notice, the one I wrote is in a goroutine and we await the actor's successful termination. in this one I think we just block the api call?

Comment (nrajanee) [reply master/internal/api_checkpoint.go:104]: Yes that makes sense.

File: master/static/migrations/20220512160917_update-job-type-jobs.tx.up.sql [hunk 23]
```diff
@@ -0,0 +1,11 @@
+ALTER TYPE public.job_type RENAME TO _job_type;
```
Comment (nrajanee) [inline master/static/migrations/20220512160917_update-job-type-jobs.tx.up.sql:1]: @stoksc there are new migrations in determined dated after this. Do I need to update the dates for these?

File: e2e_tests/tests/cluster/test_checkpoints.py [hunk 24]
```diff
@@ -215,4 +303,4 @@ def test_fail_on_preclose_chechpoint_save() -> None:
         config_obj,
         conf.fixtures_path("no_op"),
         error_log,
-    )
+    )
```
Comment (stoksc) [inline e2e_tests/tests/cluster/test_checkpoints.py:307]: don't forget to re-add this newline

File: harness/determined/cli/checkpoint.py (hunk 4 above)
Comment (stoksc) [reply harness/determined/cli/checkpoint.py:190]: let's just do commas then if experiment.py does that

File: harness/determined/cli/checkpoint.py [hunk 25]
```diff
@@ -137,6 +137,20 @@ def describe(args: Namespace) -> None:
     render_checkpoint(checkpoint)
 
 
+@authentication.required
+def delete_checkpoints(args: Namespace) -> None:
+    if args.yes or render.yes_or_no(
+        "Deleting checkpoints will result in deletion of all data associated\n"
+        "with each checkpoint in the checkpoint storage. Do you still want to proceed?"
+    ):
+        cUUIDS = args.checkpoints_uuids.split(",")
```
Comment (stoksc) [inline harness/determined/cli/checkpoint.py:146]:
```suggestion
        c_uuids = args.checkpoints_uuids.split(",")
```

File: master/internal/api_checkpoint.go (hunk 18 above)
Comment (stoksc) [reply master/internal/api_checkpoint.go:105]: sent message in dmg but I think we still need to drop the AwaitTermination() here

**Review by stoksc — APPROVED** [review]
lgtm. great work on this.

**Comment by varlogtim** [comment]
This broke `det e set gc-policy` :(

**Comment by stoksc** [comment]
https://determinedai.atlassian.net/browse/DET-8032
